### DISCUSSION TITLE
Get the sorted file list via JS API

### DISCUSSION THREAD

**Discussion by garlicbreadcleric** [SebastianMC/obsidian-custom-sort#167] [Q&A] [answered]
Hi, I'm trying to write QuickAdd macros for switching to the previous/next file in the file tree, my current script looks like this:

<details>
<summary>Code</summary>

```js
// inbox-next.js

const folderPath = "inbox/";

const getNextFile = () => {
  const files = app.vault
    .getFiles()
    .filter((file) => file.path.startsWith(folderPath));

  files.sort((a, b) => b.path.localeCompare(a.path));

  const activeFile = app.workspace.getActiveFile();
  if (activeFile == null) return files[0];

  const activeFileIndex = files.findIndex((file) => file.path === activeFile.path);

  if (activeFileIndex === -1) return files[0];

  const nextFileIndex = (activeFileIndex + 1) % files.length;
  return files[nextFileIndex];
}

module.exports = async () => {
  const nextFile = getNextFile();
  const leaf = app.workspace.getUnpinnedLeaf();
  leaf.openFile(nextFile);
};
```

</details>

This is easy to do within a specific folder (`inbox/` in my example) that has a consistent sorting, but to implement this vault-wise would require re-implementing sortspec and keeping it in sync manually. Is there some API this plugin exposes that would allow me to get the file list sorted as defined by sortspec? If not, is this something you'd consider adding?

#159 seems related, but I don't think my request requires any dependencies on internal data model of File Explorer as it's essentially just a function that accepts a file list and a sortspec and returns a new file list sorted accordingly (though I'm not at all familiar with Obsidian internals, so I might be wrong).

Thanks.

**Comment by SebastianMC** [comment, marked as answer]
Hi @garlicbreadcleric 

the relevant API function has been already there, intentionally, yet not exposed as public.

Scan `main.ts` file of the plugin for member function `orderedFolderItemsForBookmarking(...)`

The function takes as parameter a `TFolder` object, works in the context of current state of Obsidian and the custom-sort plugin to determine the sorting for the folder (standard UI-selected or custom), performs the sorting and returns ordered list of `TAbstractFile` items (that means `TFolder` and `TFile` instances)

The result of the function is exactly the same order which user sees in the File Explorer for the supplied folder.
As simple as it sounds, there are the non-obvious things:
- if the custom sorting is suspended by used, it is obviously not applied and the sort order is taken from UI (File Explorer selection) or a default is used
- if the custom sorting is in force for a folder and it depends on files metadata and/or on bookmarks order or on some other factors, these dependent data are read and used. Which means that the Obsidian metadata cache has to be populated for the function to work correctly.

Let me know if you can access the `orderedFolderItemsForBookmarking(...)` function from your plugin. This should probably work if you define the function in an interface on your side and then cast the custom-sort plugin object onto this interface.

Long term, this function can be explicitly exposed, plus a *.d.ts definition file to copy & paste. Plus some more defensive coding to make it safer for external callers.

**Reply by garlicbreadcleric** [reply]
Thanks for the response! I played around with the code a little bit yesterday, but I guess "ForBookmarking" made me assume this is something specific to bookmarks core plugin which I don't use, so I didn't look deeper into those functions.

I'll try this tomorrow and let you know if it works.

**Reply by SebastianMC** [reply]
A good point - the name of the function. 
The name is internal and specific because it reflects its only use - when a user bookmarks content of an entire folder for custom-sorting, the items are ordered via this function before passing to bookmark plugin.

Thinking in terms of future public API, the exposed function would be named in a general fashion.

BTW when you create your plugin please let me know so that I can annotate the code of custom-sort to remember about backward compatibility. Unfortunately I don't have time to do a mini-release with the renamed exposed function in the oncoming days.

**Reply by garlicbreadcleric** [reply]
Yeah that makes sense.
I need to clarify though, this isn't a public plugin I'm writing, just a user-script loaded via QuickAdd, so it's not like you changing that function would break a community plugin or anything like that.

**Reply by garlicbreadcleric** [reply]
@SebastianMC works like a charm, thanks! Here's a code example that retrieves a flat list of sorted files by recursively going through the vault folders, in case someone stumbles upon this thread looking for the same thing:

<details>
<summary>Code</summary>

```js
const getSortedFiles = () => {
  const cs = app.plugins.plugins["custom-sort"];
  if (cs == null) {
    // Implement fallback sorting here, in case custom sort plugin is disabled.
    return /* ... */;
  }

  return getSortedFolderFiles(cs, app.vault.getRoot());
};

const getSortedFolderFiles = (cs, folder) => {
  const items = cs.orderedFolderItemsForBookmarking(folder);
  return items.flatMap((item) => {
    if (item.children == null) return [item];
    return getSortedFolderFiles(cs, item);
  });
};
```

</details>
